File: Tools/Postprocessing/F_Src/fextract.f90
! extract a 1-d slice of the data (all variables or a single variable)
! along the specified coordinate direction from a plotfile.  The
! plotfile can be 1-, 2-, or 3-d.
!
! Note: by default, this routine slices through the center of the
! domain in the transverse directions.  At the moment, there is no
! runtime method of overridding this.

program fextract3d

  use bl_space
  use bl_error_module
  use bl_constants_module
  use bl_IO_module
  use plotfile_module
  use sort_d_module

  implicit none

  type(plotfile) pf
  integer :: unit, uno, un_ji
  integer :: i, j, ii, jj, kk
  integer :: rr, r1
  integer :: cnt, max_points, nvs

  real(kind=dp_t) :: dx(MAX_SPACEDIM)
  real(kind=dp_t), pointer :: p(:,:,:,:)
  real(kind=dp_t), allocatable :: sv(:,:)
  integer, allocatable :: isv(:)
  logical, allocatable :: imask(:)
  integer :: lo(MAX_SPACEDIM), hi(MAX_SPACEDIM)
  integer :: flo(MAX_SPACEDIM), fhi(MAX_SPACEDIM)
  integer :: dim
  integer :: fine_level, coarse_level

  character(len=256) :: slicefile
  character(len=256) :: pltfile
  integer :: indslsh
  integer :: iloc, jloc, kloc
  integer :: idir
  integer :: narg, farg
  character(len=256) :: fname, varnames, line, temp

  character(len=1) :: dirstr
  integer :: ivar, idx
  integer, allocatable :: var_indices(:)

  real(kind=dp_t) :: xmin, xmax, ymin, ymax, zmin, zmax

  logical :: ji_exist, valid, center
  integer :: io

  unit = unit_new()

  slicefile = ''
  pltfile  = ''
  idir = 1
  varnames = ''
  center = .true.
  coarse_level = 1
  fine_level = -1 ! This will be fixed later if the user doesn't select one

  narg = command_argument_count()

  farg = 1
  do while ( farg <= narg )
     call get_command_argument(farg, value = fname)

     select case (fname)

     case ('-p', '--pltfile')
        farg = farg + 1
        call get_command_argument(farg, value = pltfile)

     case ('-s', '--slicefile')
        farg = farg + 1
        call get_command_argument(farg, value = slicefile)

     case ('-d', '--direction')
        farg = farg + 1
        call get_command_argument(farg, value = fname)
        read(fname,*) idir

     case ('-v', '--variable')
        farg = farg + 1
        call get_command_argument(farg, value = varnames)

     case ('-l', '--lower_left')
        center = .false.

     case ('-c', '--coarse_level')
        farg = farg + 1
        call get_command_argument(farg, value = fname)
        read(fname,*) coarse_level

     case ('-f', '--fine_level')
        farg = farg + 1
        call get_command_argument(farg, value = fname)
        read(fname,*) fine_level

     case default
        exit

     end select
     farg = farg + 1
  end do

  if (len_trim(pltfile) == 0 .and. farg <= narg) then
     call get_command_argument(farg, value = pltfile)
  endif

  if ( len_trim(pltfile) == 0 ) then
     print *, " "
     print *, "Extract at 1D slice through a plotfile in any coordinate direction."
     print *, "Works with 1-, 2-, or 3-d datasets."
     print *, " "
     print *, "Usage:"
     print *, "   fextract [-p plotfile] [-s outfile] [-d dir] [-v variable] [-c coarse_level] [-f fine_level] plotfile"
     print *, " "
     print *, "args [-p|--pltfile]   plotfile        : plot file directory (depreciated, optional)"
     print *, "     [-s|--slicefile] slice file      : slice file          (optional)"
     print *, "     [-d|--direction] idir            : slice direction {1 (default), 2, or 3}"
     print *, "     [-v|--variable]  varname(s)      : only output the values of variable varname"
     print *, "                                        (space separated string for multiple variables)"
     print *, "     [-l|--lower_left]                : slice through lower left corner instead of center"
     print *, "     [-c|--coarse_level] coarse level : coarsest level to extract from"
     print *, "     [-f|--fine_level]   fine level   : finest level to extract from"
     print *, " "
     print *, "Note the plotfile at the end of the commandline is only required if you do"
     print *, "not use the deprecated '-p' option"
     print *, " "
     print *, "If a job_info file is present in the plotfile, that information is made"
     print *, "available at the end of the slice file (commented out), for reference."
     print *, " "
     stop
  end if

  ! slicefile not defined, default to plotfile.slice
  if ( len_trim(slicefile) == 0 ) then

     ! get basename of file
     indslsh = index(pltfile, '/', back = .TRUE.)

     if ( indslsh == len(trim(pltfile)) ) then
        slicefile = trim(pltfile(:indslsh-1)) // ".slice"
     else
        slicefile = trim(pltfile) // ".slice"
     end if

  endif

  select case (idir)
  case (1)
     print *, 'slicing along x-direction and outputting to ', trim(slicefile)
  case (2)
     print *, 'slicing along y-direction and outputting to ', trim(slicefile)
  case (3)
     print *, 'slicing along z-direction and outputting to ', trim(slicefile)
  end select


  call build(pf, pltfile, unit)

  nvs = pf%nvars

  ! if we are outputting only a single variable, make sure it exists
  ivar = -1
  if (varnames /= '') then
     ! flag that indicates we have variable
     ivar = 0
     allocate(var_indices(pf%nvars))
     var_indices(:) = 0

     do while (.not. trim(varnames) == "")
        ivar = ivar + 1

        idx = index(varnames, " ")
        temp = varnames(:idx)
        varnames = trim(adjustl(varnames(idx+1:)))

        var_indices(ivar) = plotfile_var_index(pf, trim(temp))
     enddo
  endif

  ! Get the index bounds and dx for the coarse level.  
  ! Note, lo and hi are 0-based indices
  lo(:) = 0
  hi(:) = 0

  lo(1:pf%dim) = lwb(plotfile_get_pd_box(pf, 1))
  hi(1:pf%dim) = upb(plotfile_get_pd_box(pf, 1))

  dx(1:pf%dim) = plotfile_get_dx(pf, 1)

  dim = pf%dim


  ! get the physical domain extrema
  xmin = pf%plo(1)
  xmax = pf%phi(1)

  if (dim >= 2) then
     ymin = pf%plo(2)
     ymax = pf%phi(2)
  endif

  if (dim == 3) then
     zmin = pf%plo(3)
     zmax = pf%phi(3)
  endif


  ! get the index bounds for the finest level
  flo(1:pf%dim) = lwb(plotfile_get_pd_box(pf, pf%flevel))
  fhi(1:pf%dim) = upb(plotfile_get_pd_box(pf, pf%flevel))

  ! compute the index of the center or lower left of the domain on the
  ! coarse grid.  These are used to set the position of the slice in
  ! the transverse direction.

  if (center) then
     iloc = (hi(1)-lo(1)+1)/2 + lo(1)
     if (dim > 1) jloc = (hi(2)-lo(2)+1)/2 + lo(2)
     if (dim > 2) kloc = (hi(3)-lo(3)+1)/2 + lo(3)
  else
     iloc = 0
     if (dim > 1) jloc = 0
     if (dim > 2) kloc = 0
  endif

  cnt = 0

  ! imask will be set to false if we've already output the data.
  ! Note, imask is defined in terms of the finest level.  As we loop
  ! over levels, we will compare to the finest level index space to
  ! determine if we've already output here

  if (idir > dim) call bl_error("idir > dim")

  select case (idir)
  case (1)
     allocate(imask(flo(1):fhi(1)))
     max_points = fhi(1) - flo(1) + 1

  case (2)
     allocate(imask(flo(2):fhi(2)))
     max_points = fhi(2) - flo(2) + 1

  case (3)
     allocate(imask(flo(3):fhi(3)))
     max_points = fhi(3) - flo(3) + 1
  end select

  imask = .true.

  ! allocate storage for the data -- we will assume that we are
  ! completely refined, and allocate enough storage for that,
  ! although, we won't really need all of this
  !
  ! note: sv(:,1) will be the coordinate information.
  ! the variables will be stored in sv(:,2:nvvs+1)
  allocate(sv(max_points,nvs+1), isv(max_points))

  ! If the user didn't select a finest level, assume we want the finest level available

  if (fine_level < 0) then
     fine_level = pf%flevel
  end if

  ! sanity check on valid selected levels

  if (fine_level > pf%flevel .or. coarse_level < 1 .or. coarse_level > fine_level) then
     call bl_error("Invalid level selection")
  end if

  ! loop over the data, starting at the finest grid, and if we haven't
  ! already store data in that grid location (according to imask),
  ! store it.  We'll put it in the correct order later.

  cnt = 0

  ! r1 is the factor between the current level grid spacing and the
  ! FINEST level
  r1  = 1

  do i = fine_level, coarse_level, -1

     ! rr is the factor between the COARSEST level grid spacing and
     ! the current level
     rr = product(pf%refrat(1:i-1,1))

     do j = 1, nboxes(pf, i)

        ! read in the data 1 patch at a time
        call fab_bind(pf, i, j)

        lo(:) = 1
        hi(:) = 1
        lo(1:pf%dim) = lwb(get_box(pf, i, j))
        hi(1:pf%dim) = upb(get_box(pf, i, j))

        p => dataptr(pf, i, j)

        select case (idir)

        case (1)

           if (dim == 1) then

              jj = 1
              kk = 1

              do ii = lo(1), hi(1)
                 if ( any(imask(ii*r1:(ii+1)*r1-1) ) ) then
                    cnt = cnt + 1

                    sv(cnt,1) = xmin + (ii + HALF)*dx(1)/rr
                    sv(cnt,2:) = p(ii,jj,kk,:)

                    imask(ii*r1:(ii+1)*r1-1) = .false.
                 end if
              end do

           else

              ! if the current patch stradles our slice, then get a data
              ! pointer to it
              if ( rr*jloc >= lo(2) .and. rr*jloc <= hi(2) .and. &
                   ( (dim .eq. 2) .or. (rr*kloc >= lo(3) .and. rr*kloc <= hi(3)) ) ) then

                 jj = jloc*rr
                 if (dim .eq. 3) then
                    kk = kloc*rr
                 else
                    kk = 1
                 end if

                 ! loop over all of the zones in the slice direction.
                 ! Here, we convert the cell-centered indices at the
                 ! current level into the corresponding RANGE on the
                 ! finest level, and test if we've stored data in any of
                 ! those locations.  If we haven't then we store this
                 ! level's data and mark that range as filled.
                 do ii = lo(1), hi(1)
                    if ( any(imask(ii*r1:(ii+1)*r1-1) ) ) then
                       cnt = cnt + 1

                       sv(cnt,1) = xmin + (ii + HALF)*dx(1)/rr
                       sv(cnt,2:) = p(ii,jj,kk,:)

                       imask(ii*r1:(ii+1)*r1-1) = .false.
                    end if
                 end do

              end if

           end if

        case (2)

           ! if the current patch stradles our slice, then get a data
           ! pointer to it
           if ( rr*iloc >= lo(1) .and. rr*iloc <= hi(1) .and. &
                ( (dim .eq. 2) .or. (rr*kloc >= lo(3) .and. rr*kloc <= hi(3)) ) ) then
              ii = iloc*rr
              if (dim .eq. 3) then
                kk = kloc*rr
              else
                kk = 1
              end if


              ! loop over all of the zones in the slice direction.
              ! Here, we convert the cell-centered indices at the
              ! current level into the corresponding RANGE on the
              ! finest level, and test if we've stored data in any of
              ! those locations.  If we haven't then we store this
              ! level's data and mark that range as filled.
              do jj = lo(2), hi(2)
                 if ( any(imask(jj*r1:(jj+1)*r1-1) ) ) then
                    cnt = cnt + 1

                    sv(cnt,1) = ymin + (jj + HALF)*dx(2)/rr
                    sv(cnt,2:) = p(ii,jj,kk,:)

                    imask(jj*r1:(jj+1)*r1-1) = .false.
                 end if
              end do

           end if

        case (3)

           ! if the current patch stradles our slice, then get a data
           ! pointer to it
           if ( rr*iloc >= lo(1) .and. rr*iloc <= hi(1) .and. &
                rr*jloc >= lo(2) .and. rr*jloc <= hi(2)) then

              ii = iloc*rr
              jj = jloc*rr

              ! loop over all of the zones in the slice direction.
              ! Here, we convert the cell-centered indices at the
              ! current level into the corresponding RANGE on the
              ! finest level, and test if we've stored data in any of
              ! those locations.  If we haven't then we store this
              ! level's data and mark that range as filled.
              do kk = lo(3), hi(3)
                 if ( any(imask(kk*r1:(kk+1)*r1-1) ) ) then
                    cnt = cnt + 1

                    sv(cnt,1) = zmin + (kk + HALF)*dx(3)/rr
                    sv(cnt,2:) = p(ii,jj,kk,:)

                    imask(kk*r1:(kk+1)*r1-1) = .false.
                 end if
              end do

           end if

        end select

        call fab_unbind(pf, i, j)
     end do

     ! adjust r1 for the next lowest level
     if ( i /= 1 ) r1 = r1*pf%refrat(i-1,1)
  end do


  ! sort the data based on the coordinates
  call sort(sv(1:cnt,1),isv(1:cnt))

  if (idir == 1) then
     dirstr = "x"
  else if (idir == 2) then
     dirstr = "y"
  else
     dirstr = "z"
  endif


 998 format("# 1-d slice in ", a1, "-direction, file: ", a)
 999 format("# time = ", g24.12)
1000 format("#",100(a24,1x))
1001 format(1x, 100(g24.12,1x))

  ! slicefile
  uno =  unit_new()
  open(unit=uno, file=slicefile, status = 'replace')

  write(uno,998) dirstr, trim(pltfile)
  write(uno,999) pf%tm

  if (ivar == -1) then

     ! output all variables

     ! write the header
     un_ji = unit_new()
     write(uno,1000) dirstr, (trim(pf%names(j)),j=1,nvs)

     ! Use this to protect against a number being xx.e-100
     !   which will print without the "e"
     do i=1,cnt
        do j=2,nvs+1
           if (abs(sv(isv(i),j)) .lt. 1.d-99) then
              sv(isv(i),j) = 0.d0
           end if
        end do
     end do

     ! write the data in columns
     do i = 1, cnt
        write(uno,1001) sv(isv(i),1), (sv(isv(i),j),j=2,nvs+1)
     end do

  else

     ! output a specific variables

     ! write the header
     write(uno,1000, advance="no") dirstr
     do i = 1, ivar
        write(uno,fmt="(a24,1x)", advance="no") trim(pf%names(var_indices(i)))
     enddo
     write(uno, *) ""

     ! Use this to protect against a number being xx.e-100
     !   which will print without the "e"
     do j=1,cnt
        do i = 1, ivar
           if (abs(sv(isv(j),var_indices(i)+1)) .lt. 1.d-99) then
              sv(isv(j),var_indices(i)+1) = 0.d0
           end if
        enddo
     enddo

     ! write the data in columns
     do j = 1, cnt
        write(uno,1001, advance="no") sv(isv(j),1)
        do i = 1, ivar
           write(uno, "(g24.12,1x)", advance="no") sv(isv(j), var_indices(i)+1)
        enddo
        write(uno, *) ""
     enddo

  endif

  ! job_info? if so write it out to the slice file end
  inquire (file=trim(pltfile)//"/job_info", exist=ji_exist)
  if (ji_exist) then
     open(unit=un_ji, file=trim(pltfile)//"/job_info", status="old")
     write (uno, fmt="(a)") " "
     valid = .true.
     do while (valid)
        read  (un_ji, fmt="(a)", iostat=io) line
        if (io < 0) then
           valid = .false.
           exit
        endif
        write (uno, fmt="('#', a)") trim(line)
     enddo

  endif

  close(unit=uno)

  call destroy(pf)

end program fextract3d

File: Tools/Plotfile/fextract.cpp
#include <AMReX.H>
#include <AMReX_Print.H>
#include <AMReX_PlotFileUtil.H>
#include <AMReX_MultiFabUtil.H>
#include <AMReX_ParallelDescriptor.H>
#include <limits>
#include <iterator>
#include <fstream>

using namespace amrex;

void main_main()
{
    const int narg = amrex::command_argument_count();

    std::string slicefile;
    std::string pltfile;
    int idir = 0;
    std::string varnames_arg;
    bool center = true;
    int coarse_level = 0;
    int fine_level = -1;  // This will be fixed later
    Real xcoord = std::numeric_limits<Real>::lowest();
    Real ycoord = std::numeric_limits<Real>::lowest();
    Real zcoord = std::numeric_limits<Real>::lowest();
    bool scientific = false;
    bool csv = false;
    int  precision = 17;
    Real tolerance = std::numeric_limits<Real>::lowest();
    bool print_info = false;

    int farg = 1;
    while (farg <= narg) {
        const std::string& name = amrex::get_command_argument(farg);
        if (name == "-s" || name == "--slicefile") {
            slicefile = amrex::get_command_argument(++farg);
        } else if (name == "-d" || name == "--direction") {
            idir = std::stoi(amrex::get_command_argument(++farg));
        } else if (name == "-v" || name == "--variable") {
            varnames_arg = amrex::get_command_argument(++farg);
        } else if (name == "-x") {
            xcoord = Real(std::stod(amrex::get_command_argument(++farg)));
        } else if (name == "-y") {
            ycoord = Real(std::stod(amrex::get_command_argument(++farg)));
        } else if (name == "-z") {
            zcoord = Real(std::stod(amrex::get_command_argument(++farg)));
        } else if (name == "-l" || name == "--lower_left") {
            center = false;
        } else if (name == "-c" || name == "--coarse_level") {
            coarse_level = std::stoi(amrex::get_command_argument(++farg));
        } else if (name == "-f" || name == "--fine_level") {
            fine_level = std::stoi(amrex::get_command_argument(++farg));
        } else if (name == "-e" || name == "--scientific") {
            scientific = true;
        } else if (name == "-csv" || name == "--csv") {
            csv = true;
        } else if (name == "-p" || name == "--precision") {
            precision = std::stoi(amrex::get_command_argument(++farg));
        } else if (name == "-t" || name == "--tolerance") {
            tolerance = Real(std::stod(amrex::get_command_argument(++farg)));
        } else if (name == "-i" || name == "--info") {
            print_info = true;
        } else {
            break;
        }
        ++farg;
    }

    if (pltfile.empty() && farg <= narg) {
        pltfile = amrex::get_command_argument(farg);
    }

    if (pltfile.empty()) {
        amrex::Print()
            << "\n"
            << " Extract at 1D slice through a plotfile in any coordinate direction.\n"
            << " Works with 1-, 2-, or 3-d datasets.\n"
            << "\n"
            << " Usage:\n"
            << "    fextract [-s outfile] [-d dir] [-v variable] [-y ycood] [-c coarse_level] [-f fine_level] plotfile\n"
            << "\n"
            << " args [-s|--slicefile] slice file      : slice file          (optional)\n"
            << "      [-d|--direction] idir            : slice direction {0 (default), 1, or 2}\n"
            << "      [-v|--variable]  varname(s)      : only output the values of variable\n"
            << "                                         varname (space separated string for\n"
            << "                                         multiple variables)\n"
            << "      [-l|--lower_left]                : slice through lower left corner\n"
            << "                                         instead of center\n"
            << "      [-x][-y][-z]                     : (x,y,z)-coordinate to pass through\n"
            << "                                         (overrides center/lower-left)\n"
            << "      [-c|--coarse_level] coarse level : coarsest level to extract from\n"
            << "      [-f|--fine_level]   fine level   : finest level to extract from\n"
            << "      [-e|--scientific]                : output data in scientific notation\n"
            << "      [-p|--precision]    precision    : decimal precision {17 (default)}\n"
            << "      [-t|--tolerance]    tolerance    : set to 0 any value lower than tolerance\n"
            << "      [-i|--info]                      : output job info\n"
            << "\n"
            << " If a job_info file is present in the plotfile, that information is made\n"
            << " available at the end of the slice file (commented out), for reference.\n"
            << std::endl;
        return;
    }

    // slicefile not defined, default to plotfile.slice
    if (slicefile.empty()) {
        slicefile = pltfile;
        if (slicefile.back() == '/') {
            slicefile.pop_back();
        }
        slicefile += ".slice";
    }

    PlotFileData pf(pltfile);
    const Vector<std::string>& var_names_pf = pf.varNames();

    Vector<std::string> var_names;
    if (varnames_arg.empty()) {
        var_names = var_names_pf;
    } else {
        std::istringstream is(varnames_arg);
        var_names.assign(std::istream_iterator<std::string>{is},
                         std::istream_iterator<std::string>{  });
        var_names.erase(std::remove_if(var_names.begin(), var_names.end(),
                                       [&](std::string const& x) {
                                           return var_names_pf.end() ==
                                               std::find(var_names_pf.begin(), var_names_pf.end(), x);
                                       }),
                        var_names.end());
        if (var_names.empty()) {
            amrex::Abort("ERROR: no valid variable names");
        }
    }

    Array<Real,AMREX_SPACEDIM> problo = pf.probLo();
    Array<Real,AMREX_SPACEDIM> dx0 = pf.cellSize(0);
    Box probdom0 = pf.probDomain(0);
    const auto lo0 = amrex::lbound(probdom0);
    const auto hi0 = amrex::ubound(probdom0);

    // compute the index of the center or lower left of the domain on the
    // coarse grid.  These are used to set the position of the slice in
    // the transverse direction.

    int iloc = 0, jloc = 0, kloc = 0;
    if (center) {
        iloc = (hi0.x-lo0.x+1)/2 + lo0.x;
        jloc = (hi0.y-lo0.y+1)/2 + lo0.y;
        kloc = (hi0.z-lo0.z+1)/2 + lo0.z;
    }

    if (xcoord > -1.e36 && AMREX_SPACEDIM >= 1) {
        // we specified the x value to pass through
        iloc = hi0.x;
        for (int i = lo0.x; i <= hi0.x; ++i) {
            amrex::Real xc = problo[0] + (Real(i)+Real(0.5))*dx0[0];
            if (xc > xcoord) {
                iloc = i;
                break;
            }
        }
    }

    if (ycoord > -1.e36 && AMREX_SPACEDIM >= 2) {
        // we specified the y value to pass through
        jloc = hi0.y;
        for (int j = lo0.y; j <= hi0.y; ++j) {
            amrex::Real yc = problo[1] + (Real(j)+Real(0.5))*dx0[1];
            if (yc > ycoord) {
                jloc = j;
                break;
            }
        }
    }

    if (zcoord > -1.e36 && AMREX_SPACEDIM == 3) {
        // we specified the z value to pass through
        kloc = hi0.z;
        for (int k = lo0.z; k <= hi0.z; ++k) {
            amrex::Real zc = problo[2] + (Real(k)+Real(0.5))*dx0[2];
            if (zc > zcoord) {
                kloc = k;
                break;
            }
        }
    }

    const int dim = pf.spaceDim();

    if (idir < 0 || idir >= dim) {
        amrex::Print() << " invalid direction\n";
        return;
    } else if (idir == 0) {
        amrex::Print() << " slicing along x-direction at coarse grid (j,k)=(" << jloc << "," << kloc << ") and output to " << slicefile << "\n";
    } else if (idir == 1) {
        amrex::Print() << " slicing along y-direction at coarse grid (i,k)=(" << iloc << "," << kloc << ") and output to " << slicefile << "\n";
    } else if (idir == 2) {
        amrex::Print() << " slicing along z-direction at coarse grid (i,j)=(" << iloc << "," << jloc << ") and output to " << slicefile << "\n";
    }


    const IntVect ivloc{AMREX_D_DECL(iloc,jloc,kloc)};

    if (fine_level < 0) fine_level = pf.finestLevel();
    // sanity check on valid selected levels
    if (fine_level > pf.finestLevel() || coarse_level < 0 || coarse_level > fine_level) {
        amrex::Abort("Invalid level selection");
    }

    Vector<Real> pos;
    Vector<Vector<Real> > data(var_names.size());

    IntVect rr{1};
    for (int ilev = coarse_level; ilev <= fine_level; ++ilev) {
        Box slice_box(ivloc*rr,ivloc*rr);
        slice_box.setSmall(idir, std::numeric_limits<int>::lowest());
        slice_box.setBig(idir, std::numeric_limits<int>::max());

        Array<Real,AMREX_SPACEDIM> dx = pf.cellSize(ilev);

        if (ilev < fine_level) {
            IntVect ratio{pf.refRatio(ilev)};
            for (int idim = dim; idim < AMREX_SPACEDIM; ++idim) {
                ratio[idim] = 1;
            }
            const iMultiFab mask = makeFineMask(pf.boxArray(ilev), pf.DistributionMap(ilev),
                                                pf.boxArray(ilev+1), ratio);
            for (int ivar = 0; ivar < var_names.size(); ++ivar) {
                const MultiFab& mf = pf.get(ilev, var_names[ivar]);
                for (MFIter mfi(mf); mfi.isValid(); ++mfi) {
                    const Box& bx = mfi.validbox() & slice_box;
                    if (bx.ok()) {
                        const auto& m = mask.array(mfi);
                        const auto& fab = mf.array(mfi);
                        const auto lo = amrex::lbound(bx);
                        const auto hi = amrex::ubound(bx);
                        for         (int k = lo.z; k <= hi.z; ++k) {
                            for     (int j = lo.y; j <= hi.y; ++j) {
                                for (int i = lo.x; i <= hi.x; ++i) {
                                    if (m(i,j,k) == 0) { // not covered by fine
                                        if (pos.size() == data[ivar].size()) {
                                            Array<Real,AMREX_SPACEDIM> p
                                                = {AMREX_D_DECL(problo[0]+static_cast<Real>(i+0.5)*dx[0],
                                                                problo[1]+static_cast<Real>(j+0.5)*dx[1],
                                                                problo[2]+static_cast<Real>(k+0.5)*dx[2])};
                                            pos.push_back(p[idir]);
                                        }
                                        data[ivar].push_back(fab(i,j,k));
                                    }
                                }
                            }
                        }
                    }
                }
            }
            rr *= ratio;
        } else {
            for (int ivar = 0; ivar < var_names.size(); ++ivar) {
                const MultiFab& mf = pf.get(ilev, var_names[ivar]);
                for (MFIter mfi(mf); mfi.isValid(); ++mfi) {
                    const Box& bx = mfi.validbox() & slice_box;
                    if (bx.ok()) {
                        const auto& fab = mf.array(mfi);
                        const auto lo = amrex::lbound(bx);
                        const auto hi = amrex::ubound(bx);
                        for         (int k = lo.z; k <= hi.z; ++k) {
                            for     (int j = lo.y; j <= hi.y; ++j) {
                                for (int i = lo.x; i <= hi.x; ++i) {
                                    if (pos.size() == data[ivar].size()) {
                                        Array<Real,AMREX_SPACEDIM> p
                                            = {AMREX_D_DECL(problo[0]+static_cast<Real>(i+0.5)*dx[0],
                                                            problo[1]+static_cast<Real>(j+0.5)*dx[1],
                                                            problo[2]+static_cast<Real>(k+0.5)*dx[2])};
                                        pos.push_back(p[idir]);
                                    }
                                    data[ivar].push_back(fab(i,j,k));
                                }
                            }
                        }
                    }
                }
            }
        }
    }

#ifdef BL_USE_MPI
    {
        const auto numpts = int(pos.size());
        auto numpts_vec = ParallelDescriptor::Gather(numpts,
                                                     ParallelDescriptor::IOProcessorNumber());
        Vector<int> recvcnt, disp;
        Vector<Real> allpos;
        Vector<Vector<Real> > alldata(data.size());
        if (ParallelDescriptor::IOProcessor()) {
            recvcnt.resize(numpts_vec.size());
            disp.resize(numpts_vec.size());
            int ntot = 0;
            disp[0] = 0;
            for (int i = 0, N = int(numpts_vec.size()); i < N; ++i) {
                ntot += numpts_vec[i];
                recvcnt[i] = numpts_vec[i];
                if (i+1 < N) {
                    disp[i+1] = disp[i] + numpts_vec[i];
                }
            }
            allpos.resize(ntot);
            alldata.resize(data.size());
            for (auto& v : alldata) {
                v.resize(ntot);
            }
        } else {
            recvcnt.resize(1);
            disp.resize(1);
            allpos.resize(1);
            for (auto& v: alldata) {
                v.resize(1);
            }
        }
        ParallelDescriptor::Gatherv(pos.data(), numpts, allpos.data(), recvcnt, disp,
                                    ParallelDescriptor::IOProcessorNumber());
        for (int i = 0; i < data.size(); ++i) {
            ParallelDescriptor::Gatherv(data[i].data(), numpts, alldata[i].data(), recvcnt, disp,
                                        ParallelDescriptor::IOProcessorNumber());
        }
        if (ParallelDescriptor::IOProcessor()) {
            pos = std::move(allpos);
            data = std::move(alldata);
        }
    }
#endif

    if (ParallelDescriptor::IOProcessor()) {
        Vector<std::pair<Real,int> > posidx;
        posidx.reserve(pos.size());
        for (int i = 0; i < pos.size(); ++i) {
            posidx.emplace_back(pos[i],i);
        }
        std::sort(posidx.begin(), posidx.end());

        const std::string dirstr = (idir == 0) ? "x" : ((idir == 1) ? "y" : "z");

        std::ofstream ofs(slicefile, std::ios::trunc);

        if (scientific) {
          ofs << std::scientific;
        }

        if (csv)
        {
           ofs << std::setw(24) << std::left << dirstr;
           for (auto const& vname : var_names) {
              ofs << ", " <<  std::setw(24) << std::left << vname;
           }
           ofs << "\n";
           for (int i = 0; i < posidx.size(); ++i) {
              ofs << std::setw(25) << std::left << std::setprecision(17) << posidx[i].first;
              for (int j = 0; j < var_names.size(); ++j) {
                  ofs << ", " << std::setw(25) << std::left << std::setprecision(17) << data[j][posidx[i].second];
              }
              ofs << "\n";
           }
        }
        else
        {
           ofs << "# 1-d slice in " << dirstr << "-direction, file: " << pltfile << "\n";
           ofs << "# time = " << std::setw(20) << std::setprecision(precision) << pf.time() << "\n";

           ofs << "#" << std::setw(24) << dirstr;
           for (auto const& vname : var_names) {
             ofs << " " <<  std::setw(24) << std::right << vname;
           }
           ofs << "\n";

           for (int i = 0; i < posidx.size(); ++i) {
             ofs << std::setw(25) << std::right << std::setprecision(precision) << posidx[i].first;
             for (int j = 0; j < var_names.size(); ++j) {
               if (std::abs(data[j][posidx[i].second])< tolerance ) data[j][posidx[i].second] = 0.;
               ofs << std::setw(25) << std::right << std::setprecision(precision) << data[j][posidx[i].second];
             }
             ofs << "\n";
           }

           if (print_info) {
             // job_info? if so write it out to the slice file end
             std::ifstream jobinfo(pltfile+"/job_info");
             if (jobinfo.good()) {
                ofs << "\n";
                std::string s;
                while (std::getline(jobinfo, s)) {
                    ofs << "#" << s << "\n";
                }
             }
           }
        }
    }
}

int main (int argc, char* argv[])
{
    amrex::SetVerbose(0);
    amrex::Initialize(argc, argv, false);
    main_main();
    amrex::Finalize();
}
